File: backend/routers/wordpress.py
"""
WordPress API Routes
REST API endpoints for WordPress integration management.
"""

from fastapi import APIRouter, HTTPException, Depends, status
from fastapi.responses import JSONResponse
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, HttpUrl
from loguru import logger

from services.integrations.wordpress_service import WordPressService
from services.integrations.wordpress_publisher import WordPressPublisher
from middleware.auth_middleware import get_current_user


router = APIRouter(prefix="/wordpress", tags=["WordPress"])


# Pydantic Models
class WordPressSiteRequest(BaseModel):
    site_url: str
    site_name: str
    username: str
    app_password: str


class WordPressSiteResponse(BaseModel):
    id: int
    site_url: str
    site_name: str
    username: str
    is_active: bool
    created_at: str
    updated_at: str


class WordPressPublishRequest(BaseModel):
    site_id: int
    title: str
    content: str
    excerpt: Optional[str] = ""
    featured_image_path: Optional[str] = None
    categories: Optional[List[str]] = None
    tags: Optional[List[str]] = None
    status: str = "draft"
    meta_description: Optional[str] = ""


class WordPressPublishResponse(BaseModel):
    success: bool
    post_id: Optional[int] = None
    post_url: Optional[str] = None
    error: Optional[str] = None


class WordPressPostResponse(BaseModel):
    id: int
    wp_post_id: int
    title: str
    status: str
    published_at: Optional[str]
    created_at: str
    site_name: str
    site_url: str


class WordPressStatusResponse(BaseModel):
    connected: bool
    sites: Optional[List[WordPressSiteResponse]] = None
    total_sites: int = 0


# Initialize services
wp_service = WordPressService()
wp_publisher = WordPressPublisher()


@router.get("/status", response_model=WordPressStatusResponse)
async def get_wordpress_status(user: dict = Depends(get_current_user)):
    """Get WordPress connection status for the current user."""
    try:
        user_id = user.get('id')
        if not user_id:
            raise HTTPException(status_code=400, detail="User ID not found")
        
        logger.info(f"Checking WordPress status for user: {user_id}")
        
        # Get user's WordPress sites
        sites = wp_service.get_all_sites(user_id)
        
        if sites:
            # Convert to response format
            site_responses = [
                WordPressSiteResponse(
                    id=site['id'],
                    site_url=site['site_url'],
                    site_name=site['site_name'],
                    username=site['username'],
                    is_active=site['is_active'],
                    created_at=site['created_at'],
                    updated_at=site['updated_at']
                )
                for site in sites
            ]
            
            logger.info(f"Found {len(sites)} WordPress sites for user {user_id}")
            return WordPressStatusResponse(
                connected=True,
                sites=site_responses,
                total_sites=len(sites)
            )
        else:
            logger.info(f"No WordPress sites found for user {user_id}")
            return WordPressStatusResponse(
                connected=False,
                sites=[],
                total_sites=0
            )
            
    except Exception as e:
        logger.error(f"Error getting WordPress status for user {user_id}: {e}")
        raise HTTPException(status_code=500, detail=f"Error checking WordPress status: {str(e)}")


@router.post("/sites", response_model=WordPressSiteResponse)
async def add_wordpress_site(
    site_request: WordPressSiteRequest,
    user: dict = Depends(get_current_user)
):
    """Add a new WordPress site connection."""
    try:
        user_id = user.get('id')
        if not user_id:
            raise HTTPException(status_code=400, detail="User ID not found")
        
        logger.info(f"Adding WordPress site for user {user_id}: {site_request.site_name}")
        
        # Add the site
        success = wp_service.add_site(
            user_id=user_id,
            site_url=site_request.site_url,
            site_name=site_request.site_name,
            username=site_request.username,
            app_password=site_request.app_password
        )
        
        if not success:
            raise HTTPException(
                status_code=400, 
                detail="Failed to connect to WordPress site. Please check your credentials."
            )
        
        # Get the added site info
        sites = wp_service.get_all_sites(user_id)
        if sites:
            latest_site = sites[0]  # Most recent site
            return WordPressSiteResponse(
                id=latest_site['id'],
                site_url=latest_site['site_url'],
                site_name=latest_site['site_name'],
                username=latest_site['username'],
                is_active=latest_site['is_active'],
                created_at=latest_site['created_at'],
                updated_at=latest_site['updated_at']
            )
        else:
            raise HTTPException(status_code=500, detail="Site added but could not retrieve details")
            
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error adding WordPress site: {e}")
        raise HTTPException(status_code=500, detail=f"Error adding WordPress site: {str(e)}")


@router.get("/sites", response_model=List[WordPressSiteResponse])
async def get_wordpress_sites(user: dict = Depends(get_current_user)):
    """Get all WordPress sites for the current user."""
    try:
        user_id = user.get('id')
        if not user_id:
            raise HTTPException(status_code=400, detail="User ID not found")
        
        logger.info(f"Getting WordPress sites for user: {user_id}")
        
        sites = wp_service.get_all_sites(user_id)
        
        site_responses = [
            WordPressSiteResponse(
                id=site['id'],
                site_url=site['site_url'],
                site_name=site['site_name'],
                username=site['username'],
                is_active=site['is_active'],
                created_at=site['created_at'],
                updated_at=site['updated_at']
            )
            for site in sites
        ]
        
        logger.info(f"Retrieved {len(sites)} WordPress sites for user {user_id}")
        return site_responses
        
    except Exception as e:
        logger.error(f"Error getting WordPress sites for user {user_id}: {e}")
        raise HTTPException(status_code=500, detail=f"Error retrieving WordPress sites: {str(e)}")


@router.delete("/sites/{site_id}")
async def disconnect_wordpress_site(
    site_id: int,
    user: dict = Depends(get_current_user)
):
    """Disconnect a WordPress site."""
    try:
        user_id = user.get('id')
        if not user_id:
            raise HTTPException(status_code=400, detail="User ID not found")
        
        logger.info(f"Disconnecting WordPress site {site_id} for user {user_id}")
        
        success = wp_service.disconnect_site(user_id, site_id)
        
        if not success:
            raise HTTPException(
                status_code=404, 
                detail="WordPress site not found or already disconnected"
            )
        
        logger.info(f"WordPress site {site_id} disconnected successfully")
        return {"success": True, "message": "WordPress site disconnected successfully"}
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error disconnecting WordPress site {site_id}: {e}")
        raise HTTPException(status_code=500, detail=f"Error disconnecting WordPress site: {str(e)}")


@router.post("/publish", response_model=WordPressPublishResponse)
async def publish_to_wordpress(
    publish_request: WordPressPublishRequest,
    user: dict = Depends(get_current_user)
):
    """Publish content to a WordPress site."""
    try:
        user_id = user.get('id')
        if not user_id:
            raise HTTPException(status_code=400, detail="User ID not found")
        
        logger.info(f"Publishing to WordPress site {publish_request.site_id} for user {user_id}")
        
        # Publish the content
        result = wp_publisher.publish_blog_post(
            user_id=user_id,
            site_id=publish_request.site_id,
            title=publish_request.title,
            content=publish_request.content,
            excerpt=publish_request.excerpt,
            featured_image_path=publish_request.featured_image_path,
            categories=publish_request.categories,
            tags=publish_request.tags,
            status=publish_request.status,
            meta_description=publish_request.meta_description
        )
        
        if result['success']:
            logger.info(f"Content published successfully to WordPress: {result['post_id']}")
            return WordPressPublishResponse(
                success=True,
                post_id=result['post_id'],
                post_url=result.get('post_url')
            )
        else:
            logger.error(f"Failed to publish content: {result['error']}")
            return WordPressPublishResponse(
                success=False,
                error=result['error']
            )
            
    except Exception as e:
        logger.error(f"Error publishing to WordPress: {e}")
        return WordPressPublishResponse(
            success=False,
            error=f"Error publishing content: {str(e)}"
        )


@router.get("/posts", response_model=List[WordPressPostResponse])
async def get_wordpress_posts(
    site_id: Optional[int] = None,
    user: dict = Depends(get_current_user)
):
    """Get published posts from WordPress sites."""
    try:
        user_id = user.get('id')
        if not user_id:
            raise HTTPException(status_code=400, detail="User ID not found")
        
        logger.info(f"Getting WordPress posts for user {user_id}, site_id: {site_id}")
        
        posts = wp_service.get_posts_for_site(user_id, site_id) if site_id else wp_service.get_posts_for_all_sites(user_id)
        
        post_responses = [
            WordPressPostResponse(
                id=post['id'],
                wp_post_id=post['wp_post_id'],
                title=post['title'],
                status=post['status'],
                published_at=post['published_at'],
                created_at=post['created_at'],
                site_name=post['site_name'],
                site_url=post['site_url']
            )
            for post in posts
        ]
        
        logger.info(f"Retrieved {len(posts)} WordPress posts for user {user_id}")
        return post_responses
        
    except Exception as e:
        logger.error(f"Error getting WordPress posts for user {user_id}: {e}")
        raise HTTPException(status_code=500, detail=f"Error retrieving WordPress posts: {str(e)}")


@router.put("/posts/{post_id}/status")
async def update_post_status(
    post_id: int,
    status: str,
    user: dict = Depends(get_current_user)
):
    """Update the status of a WordPress post (draft/publish)."""
    try:
        user_id = user.get('id')
        if not user_id:
            raise HTTPException(status_code=400, detail="User ID not found")
        
        if status not in ['draft', 'publish', 'private']:
            raise HTTPException(
                status_code=400, 
                detail="Invalid status. Must be 'draft', 'publish', or 'private'"
            )
        
        logger.info(f"Updating WordPress post {post_id} status to {status} for user {user_id}")
        
        success = wp_publisher.update_post_status(user_id, post_id, status)
        
        if not success:
            raise HTTPException(
                status_code=404, 
                detail="Post not found or update failed"
            )
        
        logger.info(f"WordPress post {post_id} status updated to {status}")
        return {"success": True, "message": f"Post status updated to {status}"}
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error updating WordPress post {post_id} status: {e}")
        raise HTTPException(status_code=500, detail=f"Error updating post status: {str(e)}")


@router.delete("/posts/{post_id}")
async def delete_wordpress_post(
    post_id: int,
    force: bool = False,
    user: dict = Depends(get_current_user)
):
    """Delete a WordPress post."""
    try:
        user_id = user.get('id')
        if not user_id:
            raise HTTPException(status_code=400, detail="User ID not found")
        
        logger.info(f"Deleting WordPress post {post_id} for user {user_id}, force: {force}")
        
        success = wp_publisher.delete_post(user_id, post_id, force)
        
        if not success:
            raise HTTPException(
                status_code=404, 
                detail="Post not found or deletion failed"
            )
        
        logger.info(f"WordPress post {post_id} deleted successfully")
        return {"success": True, "message": "Post deleted successfully"}
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting WordPress post {post_id}: {e}")
        raise HTTPException(status_code=500, detail=f"Error deleting post: {str(e)}")


@router.get("/health")
async def wordpress_health_check():
    """WordPress integration health check."""
    try:
        return {
            "status": "healthy",
            "service": "wordpress",
            "timestamp": "2024-01-01T00:00:00Z",
            "version": "1.0.0"
        }
    except Exception as e:
        logger.error(f"WordPress health check failed: {e}")
        raise HTTPException(status_code=500, detail="WordPress service unhealthy")

File: frontend/src/api/wordpress.ts
/**
 * WordPress API client for ALwrity frontend.
 * Handles WordPress site connections, content publishing, and management.
 */

import { apiClient } from './client';

export interface WordPressSite {
  id: number;
  site_url: string;
  site_name: string;
  username: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface WordPressSiteRequest {
  site_url: string;
  site_name: string;
  username: string;
  app_password: string;
}

export interface WordPressPublishRequest {
  site_id: number;
  title: string;
  content: string;
  excerpt?: string;
  featured_image_path?: string;
  categories?: string[];
  tags?: string[];
  status?: 'draft' | 'publish' | 'private';
  meta_description?: string;
}

export interface WordPressPublishResponse {
  success: boolean;
  post_id?: number;
  post_url?: string;
  error?: string;
}

export interface WordPressPost {
  id: number;
  wp_post_id: number;
  title: string;
  status: string;
  published_at?: string;
  created_at: string;
  site_name: string;
  site_url: string;
}

export interface WordPressStatusResponse {
  connected: boolean;
  sites?: WordPressSite[];
  total_sites: number;
}

export interface WordPressHealthResponse {
  status: string;
  service: string;
  timestamp: string;
  version: string;
}

class WordPressAPI {
  private baseUrl = '/wordpress';
  private getAuthToken: (() => Promise<string | null>) | null = null;

  /**
   * Set authentication token getter
   */
  setAuthTokenGetter(getter: () => Promise<string | null>) {
    this.getAuthToken = getter;
  }

  /**
   * Get authenticated client with token
   */
  private async getAuthenticatedClient() {
    if (this.getAuthToken) {
      const token = await this.getAuthToken();
      if (token) {
        // Create a new client instance with the auth header
        return apiClient.create({
          headers: {
            'Authorization': `Bearer ${token}`
          }
        });
      }
    }
    return apiClient;
  }

  /**
   * Get WordPress connection status
   */
  async getStatus(): Promise<WordPressStatusResponse> {
    try {
      const client = await this.getAuthenticatedClient();
      const response = await client.get(`${this.baseUrl}/status`);
      return response.data;
    } catch (error) {
      console.error('WordPress API: Error getting status:', error);
      throw error;
    }
  }

  /**
   * Add a new WordPress site connection
   */
  async addSite(siteData: WordPressSiteRequest): Promise<WordPressSite> {
    try {
      const client = await this.getAuthenticatedClient();
      const response = await client.post(`${this.baseUrl}/sites`, siteData);
      return response.data;
    } catch (error) {
      console.error('WordPress API: Error adding site:', error);
      throw error;
    }
  }

  /**
   * Get all WordPress sites for the current user
   */
  async getSites(): Promise<WordPressSite[]> {
    try {
      const client = await this.getAuthenticatedClient();
      const response = await client.get(`${this.baseUrl}/sites`);
      return response.data;
    } catch (error) {
      console.error('WordPress API: Error getting sites:', error);
      throw error;
    }
  }

  /**
   * Disconnect a WordPress site
   */
  async disconnectSite(siteId: number): Promise<{ success: boolean; message: string }> {
    try {
      const client = await this.getAuthenticatedClient();
      const response = await client.delete(`${this.baseUrl}/sites/${siteId}`);
      return response.data;
    } catch (error) {
      console.error('WordPress API: Error disconnecting site:', error);
      throw error;
    }
  }

  /**
   * Publish content to WordPress
   */
  async publishContent(publishData: WordPressPublishRequest): Promise<WordPressPublishResponse> {
    try {
      const client = await this.getAuthenticatedClient();
      const response = await client.post(`${this.baseUrl}/publish`, publishData);
      return response.data;
    } catch (error) {
      console.error('WordPress API: Error publishing content:', error);
      throw error;
    }
  }

  /**
   * Get published posts from WordPress sites
   */
  async getPosts(siteId?: number): Promise<WordPressPost[]> {
    try {
      const client = await this.getAuthenticatedClient();
      const params = siteId ? { site_id: siteId } : {};
      const response = await client.get(`${this.baseUrl}/posts`, { params });
      return response.data;
    } catch (error) {
      console.error('WordPress API: Error getting posts:', error);
      throw error;
    }
  }

  /**
   * Update post status (draft/publish/private)
   */
  async updatePostStatus(postId: number, status: string): Promise<{ success: boolean; message: string }> {
    try {
      const client = await this.getAuthenticatedClient();
      const response = await client.put(`${this.baseUrl}/posts/${postId}/status`, null, {
        params: { status }
      });
      return response.data;
    } catch (error) {
      console.error('WordPress API: Error updating post status:', error);
      throw error;
    }
  }

  /**
   * Delete a WordPress post
   */
  async deletePost(postId: number, force: boolean = false): Promise<{ success: boolean; message: string }> {
    try {
      const client = await this.getAuthenticatedClient();
      const response = await client.delete(`${this.baseUrl}/posts/${postId}`, {
        params: { force }
      });
      return response.data;
    } catch (error) {
      console.error('WordPress API: Error deleting post:', error);
      throw error;
    }
  }

  /**
   * Test WordPress site connection
   */
  async testConnection(siteData: WordPressSiteRequest): Promise<boolean> {
    try {
      // This would typically be a separate endpoint for testing connections
      // For now, we'll try to add the site and see if it succeeds
      await this.addSite(siteData);
      return true;
    } catch (error) {
      console.error('WordPress API: Connection test failed:', error);
      return false;
    }
  }

  /**
   * Health check
   */
  async healthCheck(): Promise<WordPressHealthResponse> {
    try {
      const response = await apiClient.get(`${this.baseUrl}/health`);
      return response.data;
    } catch (error) {
      console.error('WordPress API: Health check failed:', error);
      throw error;
    }
  }

  /**
   * Validate WordPress site URL
   */
  validateSiteUrl(url: string): boolean {
    try {
      // Remove protocol if present
      const cleanUrl = url.replace(/^https?:\/\//, '');
      
      // Basic URL validation
      const urlPattern = /^[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]*\.([a-zA-Z]{2,}|[a-zA-Z]{2,}\.[a-zA-Z]{2,})$/;
      
      return urlPattern.test(cleanUrl) || cleanUrl.includes('localhost') || cleanUrl.includes('127.0.0.1');
    } catch (error) {
      return false;
    }
  }

  /**
   * Format WordPress site URL
   */
  formatSiteUrl(url: string): string {
    if (!url) return '';
    
    // Add protocol if missing
    if (!url.startsWith('http://') && !url.startsWith('https://')) {
      return `https://${url}`;
    }
    
    return url;
  }
}

// Export singleton instance
export const wordpressAPI = new WordPressAPI();
export default wordpressAPI;
